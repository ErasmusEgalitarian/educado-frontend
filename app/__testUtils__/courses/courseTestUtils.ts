import { Course, Section } from "../../src/interfaces/Course";

interface PartialCourse {
  _id?: string;
  title?: string;
  category?: string;
  coverImg?: string;
  description?: string;
  creator?: string;
  difficulty?: number;
  status?: string;
  estimatedHours?: number;
  rating?: number;
  numOfSubscriptions?: number;
  dateUpdated?: Date;
  dateCreated?: Date;
  sections?: Section[];
  __v?: number;
}

/**
 * Generates a mock course for testing from partial options
 *
 * @param options an object containing just the course options you want to set
 * @returns Course
 */
export const generateTestCourse = (options: PartialCourse = {}): Course => {
  return {
    _id: options._id ?? "1",
    title: options.title ?? "Test course",
    description: options.description ?? "A mock course for testing",
    dateCreated: options.dateCreated ?? new Date("2023-10-10T00:00:00.000Z"),
    dateUpdated: options.dateUpdated ?? new Date("2023-10-13T13:01:00.000Z"),
    coverImg: options.coverImg ?? "",
    category: options.category ?? "health and workplace safety",
    sections: options.sections ?? [],
    creator: options.creator ?? "",
    difficulty: options.difficulty ?? 2,
    status: options.status ?? "draft",
    estimatedHours: options.estimatedHours ?? 8.5,
    rating: options.rating ?? 4.5,
    numOfSubscriptions: options.numOfSubscriptions ?? 157,
    __v: options.__v ?? 0,
  };
};
