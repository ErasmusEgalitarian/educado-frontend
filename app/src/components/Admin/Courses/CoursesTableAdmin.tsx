/* eslint-disable @typescript-eslint/strict-boolean-expressions */
import { useEffect, useState } from "react";
import { IconContext } from "react-icons";
import { MdStar } from "react-icons/md";
import useSWR from "swr";

import { CoursesDeleteButton } from "@components/Admin/Courses/Actions/CoursesDeleteButton";
import { CoursesUpdateButton } from "@components/Admin/Courses/Actions/CoursesUpdateButton";
import { SearchBar } from "@components/SearchBar/SearchBar";
import {
  Table,
  TableHead,
  TableBody,
  TableRow,
  TableCell,
  TableContainer,
  PaginationBottomBar,
} from "@components/Table";
import { usePagination } from "@hooks/usePagination";
import { CreatorPopulatedCourse } from "@interfaces/Course";
import courseService from "@services/course.services";

export const CoursesTableAdmin = () => {
  const [filteredCourses, setFilteredCourses] = useState<
    CreatorPopulatedCourse[]
  >([]);

  const { data: coursesResponse, mutate } = useSWR("api/courses", () =>
    courseService.getAllCourses(),
  );

  useEffect(() => {
    if (coursesResponse !== undefined) setFilteredCourses(coursesResponse);
  }, [coursesResponse]);

  //this should be generalized more and turned into a hook
  const searchFunction = (searchString: string) => {
    if (!coursesResponse) return;

    const filteredItems = coursesResponse.filter((course) => {
      if (searchString === "") return course;

      // this will be typed in a better way when a hook is made
      const fieldsToCheck = [
        "title",
        "creator",
        "category",
        "numOfSubscriptions",
        "rating",
      ] as const;

      const user = course.creator?.baseUser;
      if (user) {
        const nameString = `${user.firstName} ${user.lastName}`;
        return nameString.toLowerCase().includes(searchString.toLowerCase());
      }

      return fieldsToCheck.some((field) => {
        const valueToSearch = course[field];

        if (valueToSearch === null || valueToSearch === undefined) return false;

        switch (typeof valueToSearch) {
          case "string":
            return valueToSearch
              .toLowerCase()
              .includes(searchString.toLowerCase());
          case "number":
            return valueToSearch === Number(searchString);
          default:
            return false;
        }
      });
    });
    handleResetPagination();
    setFilteredCourses(filteredItems);
  };

  const {
    isPaginated,
    paginatedItems: paginatedCourses,
    currentPage,
    itemsPerPage,
    handleChangePage,
    handleChangeItemsPerPage,
    handleResetPagination,
  } = usePagination(filteredCourses, 10);

  const rows = paginatedCourses.map((course, key) => {
    const fullName =
      course.creator && course.creator.baseUser
        ? `${course.creator.baseUser.firstName} ${course.creator.baseUser.lastName}`
        : null;

    return (
      //find a nice way to handle react keys
      <TableRow
        key={`${course.title}-${key}`}
        cells={[
          <TableCell>
            <p className="whitespace-normal">{course.title}</p>
          </TableCell>,
          <TableCell>
            {fullName && <p className="whitespace-normal">{fullName}</p>}
          </TableCell>,
          <TableCell>
            {course.category && (
              <p className="whitespace-normal">
                {course.category.charAt(0).toLocaleUpperCase() +
                  course.category.slice(1)}
              </p>
            )}
          </TableCell>,
          <TableCell>
            {course.numOfSubscriptions && (
              <p className="whitespace-normal">{`${course.numOfSubscriptions} alunos`}</p>
            )}
          </TableCell>,
          <TableCell>
            <div className="flex space-x-1">
              <MdStar className="text-yellow-400 self-center" />
              <p className="text-yellow-400 text-lg font-bold">
                {course.rating?.toFixed(1)}
              </p>
            </div>
          </TableCell>,
          <TableCell>
            <div className="flex flex-wrap justify-end gap-2">
              <div>
                <CoursesUpdateButton course={course} refreshFn={mutate} />
              </div>
              <div>
                <CoursesDeleteButton courseId={course._id} refreshFn={mutate} />
              </div>
            </div>
          </TableCell>,
        ]}
      />
    );
  });

  return (
    <TableContainer>
      <SearchBar
        sortingOptions={[
          { displayName: "Mais recentes", htmlValue: "most-recent" },
        ]}
        placeholderText="Buscar Cursos"
        searchFn={searchFunction}
      />
      <Table>
        <TableHead>
          <TableRow
            isHeaderRow
            cells={[
              <TableCell key="name">
                <p>Nome</p>
              </TableCell>,
              <TableCell key="creator">
                <p>Criador de Conte√∫do</p>
              </TableCell>,
              <TableCell key="category">
                <p>Categoria</p>
              </TableCell>,
              <TableCell key="subscribers">
                <p>Inscritos</p>
              </TableCell>,
              <TableCell key="rating">
                <p>Nota</p>
              </TableCell>,
              <TableCell key="actions">
                {/* Empty col for actions */}
              </TableCell>,
            ]}
          />
        </TableHead>
        <TableBody>
          {/* Icon context to give all the react-icons in the rows the same options */}
          <IconContext.Provider value={{ size: "20" }}>
            {rows}
          </IconContext.Provider>
        </TableBody>
      </Table>
      {coursesResponse && isPaginated && (
        <PaginationBottomBar
          currentPage={currentPage}
          itemsPerPage={itemsPerPage}
          unpaginatedItemsAmount={filteredCourses.length}
          onChangePage={handleChangePage}
          onChangeItemsPerPage={handleChangeItemsPerPage}
        />
      )}
    </TableContainer>
  );
};
