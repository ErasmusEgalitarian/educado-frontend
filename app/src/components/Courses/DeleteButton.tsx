import { useState } from "react";
import { useNavigate } from "react-router-dom";
import { toast } from "react-toastify";

// Services
import CourseServices from "../../services/course.services";
import SectionServices from "../../services/section.services";
import { useNotifications } from "../notification/NotificationContext";

export enum DeleteType {
  course = "course",
  section = "section",
  lecture = "lecture",
  exercise = "exercise",
}

interface DeleteButtonProps {
  id: string | undefined;
  token: string;
  deleteType: DeleteType;
}

export const DeleteButton = ({ id, token, deleteType }: DeleteButtonProps) => {
  const [res, setRes] = useState<any>();

  const navigate = useNavigate();
  const { addNotification } = useNotifications();

  /**
   * Delete courses and redirect to courses page
   * Uses window.location.href to redirect instead of navigate, as navigate doesn't update the page
   *
   * @param id The course id
   * @param token The user token
   */
  const clickDelete = async () => {
    if (confirm("VocÃª tem certeza?")) {
      // Checks what type of delete is being made
      switch (deleteType) {
        case "course":
          deleteCourse(id, token);
          break;

        case "section":
          deleteSection(id, token);
          break;

        case "lecture":
          break;

        case "exercise":
          break;

        default:
          break;
      }
    }
  };

  /**
   * Delete the course and redirect to courses page
   *
   * @param id Course id
   * @param token User token
   */
  const deleteCourse = async (id: string | undefined, token: string) => {
    setRes(await CourseServices.deleteCourse(id, token));

    // const responseFile = await StorageService.deleteFile(ImgId, token);

    if (statusFeedback()) {
      navigate("/courses");
    }
  };

  /**
   * Delete the Section and redirect to the parent course page
   *
   * @param id Section id
   * @param token User token
   */
  const deleteSection = async (id: string | undefined, token: string) => {
    setRes(await SectionServices.deleteSection(id, token));

    if (statusFeedback()) {
      navigate("/courses"); // Should be the parent course page
    }
  };

  /**
   * Display a toast message based on the response status code
   *
   * @param status The status code of the response
   */
  const statusFeedback = () => {
    if (res.status >= 200 && res.status <= 299) {
      addNotification("ExcluÃ­do");
      {
        /* Course deleted */
      }
      return true;
    } else if (res.status >= 400 && res.status <= 599) {
      toast.error(
        `(${res.status}, ${res.statusText}) while attempting to delete course`,
      );
      return false;
    }
  };

  return (
    <button
      type="button"
      onClick={clickDelete}
      className="left-0 std-button bg-warning hover:bg-red-800 ml-4"
    >
      Excluir
    </button>
  );
};
