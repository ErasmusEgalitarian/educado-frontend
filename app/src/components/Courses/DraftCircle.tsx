// Interfaces
import statuses from "../../helpers/courseStatuses";

interface DraftCircleProps {
  statusSTR: string;
}

export const DraftCircle = ({ statusSTR }: DraftCircleProps) => {
  return (
    <div>
      <div
        className={
          "w-3 h-3 mx-2 rounded-full m-auto " +
          (statuses[statusSTR].color ?? statuses.default.color)
        }
      />
      <p className="italic">{statuses[statusSTR].br ?? statuses.default.br}</p>
    </div>
  );
};
