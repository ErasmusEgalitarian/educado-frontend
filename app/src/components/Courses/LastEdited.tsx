// Interfaces
import { mdiPencil } from "@mdi/js";
import { Icon } from "@mdi/react";

import { Course } from "../../interfaces/Course";

// Components

/**
 * Returns a string with the time difference between two dates.
 * > Note: doesn't account for leap years, different month lengths, etc.\
 * > Couldn't be arsed, I'm sorry
 *
 * @param fromDate
 * @param toDate
 * @returns String
 */
const getFormattedTimeDifference = (fromDate: Date, toDate: Date): string => {
  const minuteDifference = Math.floor(
    (toDate.getTime() - fromDate.getTime()) / 1000 / 60,
  );

  // Minutes
  if (minuteDifference < 60) {
    return `${minuteDifference} minutos atrás`;
  }
  // Hours
  if (minuteDifference < 60 * 24) {
    return `${Math.floor(minuteDifference / 60)} horas atrás`;
  }
  // Days
  if (minuteDifference < 60 * 24 * 7) {
    return `${Math.floor(minuteDifference / 60 / 24)} dias atrás`;
  }
  // Weeks
  if (minuteDifference < 60 * 24 * 30) {
    return `${Math.floor(minuteDifference / 60 / 24)} semanas atrás`;
  }
  // Months
  if (minuteDifference < 60 * 24 * 365) {
    return `${Math.floor(minuteDifference / 60 / 24 / 30)} meses atrás`;
  }

  // Show date if difference is > 1 year
  return fromDate.toLocaleDateString();
};

/**
 * Displays the last time a course was edited
 *
 * @param course
 * @returns HTML Element
 */
export const LastEdited = ({ course }: { course: Course }) => {
  return (
    <>
      <div className="flex flex-row">
        <Icon path={mdiPencil} className="w-4 mr-1 inline-block mr-1" />
        <p className="text-md italic">
          {getFormattedTimeDifference(
            new Date(course.dateUpdated ?? Date.now()),
            new Date(),
          )}
        </p>
      </div>
    </>
  );
};
