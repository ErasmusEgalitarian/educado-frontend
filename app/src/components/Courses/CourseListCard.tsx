// interfaces
import { mdiPencil, mdiAccount } from "@mdi/js";
import { Icon } from "@mdi/react";
import { useEffect, useState } from "react";

import imageNotFoundImage from "../../assets/image-not-found.png";
import categories from "../../helpers/courseCategories";
import statuses from "../../helpers/courseStatuses";
import { Course } from "../../interfaces/Course";

// Components
import {
  getAverageRatingOfCC,
  getAverageRatingOfCourse,
} from "../../services/contentCreator.services";
import StorageServices from "../../services/storage.services";
import StarRating from "../general/StarRating";

import { LastEdited } from "./LastEdited";

// Helpers

// Images

// Rating

/**
 * Displays a course in a card format
 *
 * @param {Course} course The course to be displayed
 * @returns HTML Element
 */
export const CourseListCard = ({ course }: { course: Course }) => {
  const [imageSrc, setImageSrc] = useState<string>("");
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const maxTitleLength = 20;
  //Only load the picture, when the picture is loaded
  useEffect(() => {
    const fetchImage = async () => {
      try {
        if (course.coverImg == "" || course.coverImg == undefined) {
          throw new Error("coverImg is empty or undefined");
        }
        const fileSrc = await StorageServices.getMedia(course.coverImg);

        setImageSrc(fileSrc);
      } catch (error) {
        setImageSrc(imageNotFoundImage);
      } finally {
        setIsLoading(false);
      }
    };
    const fetchRating = async () => {
      try {
        const rating = await getAverageRatingOfCourse(course._id);
        console.log("card" + rating);
      } catch (error) {
        console.log(error);
      }
    };

    fetchRating();
    fetchImage();
  }, [course.coverImg, course.title]);
  return (
    <div className="overflow-hidden shadow rounded h-full w-full cursor-pointer m-auto hover:shadow-lg duration-200">
      <label
        onClick={() =>
          (window.location.href = `/courses/manager/${course._id}/0`)
        }
        className="w-full block h-full relative group"
      >
        {/* Hover overlay */}
        <div className="absolute inset-0 grid place-content-center">
          {/* Edit button */}
          <div className="z-10 bg-primary opacity-0 rounded-full p-8 group-hover:animate-slidePopIn duration-200 group-hover:opacity-100">
            <Icon path={mdiPencil} className="w-12 text-white" />
          </div>
          <div className="z-0 absolute inset-0 bg-white opacity-0 group-hover:opacity-30 duration-200" />
        </div>

        {/* Card image */}
        {isLoading ? (
          <div className="h-40 w-full bg-gray-200 animate-pulse" />
        ) : (
          <img
            src={imageSrc}
            className="h-40 w-full object-cover bg-white border-b"
          />
        )}

        {/* Card content */}
        <div className="bg-white w-full">
          <div className="p-4">
            {/* Card title */}
            <h2 className="text-gray-800 text-xl font-medium mb-2">
              {course.title.slice(0, maxTitleLength) +
                (course.title.length > maxTitleLength ? "..." : "")}
            </h2>

            <div className="flex flex-row justify-between">
              {/* Course rating */}
              <div className="w-[8rem]">
                <StarRating rating={course.rating ? course.rating : 0} />
              </div>
              {/* Subsriber count */}
              <div className="flex flex-row">
                <Icon
                  path={mdiAccount}
                  className="w-5 mr-1 inline-block mr-1"
                />
                <p className="text-grayMedium">{course.numOfSubscriptions}</p>
              </div>
            </div>

            {/* Card info */}
            <div className="h-20 flex flex-col justify-between">
              {/* Course category */}
              <div className="flex flex-row">
                <Icon
                  path={
                    course.category
                      ? categories[course.category]?.icon
                      : categories.default.icon
                  }
                  className="w-4 mr-1"
                />
                <p>{categories[course.category]?.br ?? course.category}</p>
              </div>

              <div className="flex flex-row justify-between">
                {/* Course last modified */}
                <div>
                  <LastEdited course={course} />
                </div>

                {/* Course status */}
                <div className="flex flex-row">
                  <div
                    className={
                      "w-3 h-3 mx-2 rounded-full m-auto " +
                      (course.status
                        ? statuses[course.status].color
                        : statuses.default.color)
                    }
                  />
                  <p className="italic">
                    {course.status
                      ? statuses[course.status].br
                      : statuses.default.br}
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </label>
    </div>
  );
};
