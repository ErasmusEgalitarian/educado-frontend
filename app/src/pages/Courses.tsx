// Hooks
import {
  InformationCircleIcon,
  PencilSquareIcon,
  QuestionMarkCircleIcon,
} from "@heroicons/react/24/outline";
import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { toast } from "react-toastify";
import useSWR from "swr";

import CourseServices from "@services/course.services";

import noCoursesImage from "../assets/no-courses.png";
import { CourseListCard } from "../components/Courses/CourseListCard";
import CourseGuideButton from "../components/Courses/GuideToCreatingCourse";

// Services

// Components
import PersonalInsights from "../components/Courses/PersonalInsights";
import Loading from "../components/general/Loading";
import Layout from "../components/Layout";

// toasts

// Images
import { getUserToken } from "../helpers/userInfo";

/**
 * @returns HTML Element
 *
 * This page displays all courses created by the user\
 * as well as some personal insights about the user.
 */
const Courses = () => {
  // States and Hooks
  const navigate = useNavigate();
  const token = getUserToken();
  // Fetch all courses

  // TODO: Implement proper backend call once backend is ready

  const CourseManager = () => {
    navigate("/courses/manager/0/0");
  };

  const [showTutorial, setShowTutorial] = useState(false);

  const handleShowTutorial = () => {
    setShowTutorial(true);
  };

  const handleCloseTutorial = () => {
    setShowTutorial(false);
  };

  const { data, error } = useSWR(
    token ? [token] : null,
    CourseServices.getAllCreatorCourses,
  );

  if (error && error.response.status === 401) {
    localStorage.removeItem("token");
    localStorage.removeItem("userInfo");
    navigate("/login");
    return null;
  }

  if (!data)
    return (
      <Layout meta="course overview">
        <Loading />
      </Layout>
    );

  return (
    <Layout meta="Course overview">
      <div className="grid lg:grid-cols-[3fr_1fr] h-full font-personalInsights">
        {/* Left side displaying courses, filtering for these and create new button */}
        <div className="m-8 p-8 pb-0 bg-white rounded-xl overflow-hidden flex flex-col">
          {data.length ? (
            <>
              {/* Header and create course button */}
              <div className="flex flex-row no-wrap">
                <h1 className="text-3xl font-bold flex-1">
                  Confira seus cursos
                </h1>
                <div className="flex flex-row gap-5">
                  <label
                    htmlFor="course-create"
                    onClick={CourseManager}
                    className="std-button flex modal-button space-x-2 cursor-pointer"
                  >
                    <PencilSquareIcon className="w-8 h-8" />
                    <p className="font-normal ">Criar novo curso</p>{" "}
                    {/** Create new course */}
                  </label>
                  {/* Course guide button shows a tutorial*/}
                  <CourseGuideButton />
                </div>
              </div>
              {/* Card/compact view toggle and filters */}
              <div className="h-10 my-8 bg-grayLight">
                {/* TODO: Implement card/compact view toggle */}

                {/* TODO: Implement filters */}
              </div>
              <div className="overflow-y-scroll no-scrollbar">
                <div className="grid grid-cols-1 md:grid-cols-2 xl:grid-cols-3 3xl:grid-cols-4 gap-6 pb-4">
                  {data.map((course: any, key: number) => (
                    <CourseListCard course={course} key={key} />
                  ))}
                </div>
              </div>
            </>
          ) : (
            // If the user has no courses, display this 'empty state'
            <div className="grid place-content-center w-full h-full text-center">
              <div className="md:mx-40 xl:mx-64">
                <img src={noCoursesImage} className="w-full" />
                <h1 className="text-xl font-bold my-4">Comece agora</h1>
                {/* You haven't created any courses yet.
              Click the button below and follow the
              step-by-step instructions to develop your first course. */}
                <p>
                  Você ainda não criou nenhum curso. Clique no botão abaixo e
                  siga o passo a passo para desenolver o seu primeiro curso.
                </p>
                {/* Create course button */}
                <label
                  htmlFor="course-create"
                  onClick={CourseManager}
                  className="std-button flex modal-button  space-x-2"
                >
                  <PencilSquareIcon className="w-5 h-5" />
                  <p className="font-normal ">Criar novo curso</p>{" "}
                  {/** Create new course */}
                </label>
              </div>
            </div>
          )}
        </div>
        {/* Right side displaying personal insights */}
        <div className="m-8 hidden lg:block">
          <PersonalInsights courses={data} />
        </div>
      </div>
    </Layout>
  );
};

export default Courses;
