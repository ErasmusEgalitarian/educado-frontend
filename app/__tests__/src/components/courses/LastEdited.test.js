import renderer from "react-test-renderer";
import { describe, it, expect } from "@jest/globals";
import { generateTestCourse } from "../../../../__testUtils__/courses/courseTestUtils";

// Components
import { LastEdited } from "../../../../src/components/Courses/LastEdited";
import { MemoryRouter } from "react-router-dom";

describe("last edited time", () => {
  it("shows minutes if difference is below 1 hour", () => {
    // Set lastModified to 59 minutes ago
    const course = generateTestCourse({
      dateUpdated: new Date(Date.now() - 60 * 60 * 1000 + 1000),
    });
    const component = renderer.create(
      <MemoryRouter>
        <LastEdited course={course} key={1} />
      </MemoryRouter>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows hours if difference is below 24 hours", () => {
    // Set lastModified to 23 hours and 59 minutes ago
    const course = generateTestCourse({
      dateUpdated: new Date(Date.now() - 24 * 60 * 60 * 1000 + 1000),
    });
    const component = renderer.create(
      <MemoryRouter>
        <LastEdited course={course} key={1} />
      </MemoryRouter>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows days if difference is below 1 week", () => {
    // Set lastModified to 6 days ago
    const course = generateTestCourse({
      dateUpdated: new Date(Date.now() - 7 * 24 * 60 * 60 * 1000 + 1000),
    });
    const component = renderer.create(
      <MemoryRouter>
        <LastEdited course={course} key={1} />
      </MemoryRouter>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows weeks if difference is below 30 days", () => {
    // Set lastModified to 29 days ago
    const course = generateTestCourse({
      dateUpdated: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000 + 1000),
    });
    const component = renderer.create(
      <MemoryRouter>
        <LastEdited course={course} key={1} />
      </MemoryRouter>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows months if difference is below 1 year", () => {
    // Set lastModified to 364 days ago
    const course = generateTestCourse({
      dateUpdated: new Date(Date.now() - 365 * 24 * 60 * 60 * 1000 + 1000),
    });
    const component = renderer.create(
      <MemoryRouter>
        <LastEdited course={course} key={1} />
      </MemoryRouter>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows fromDate if difference is above 1 year", () => {
    const testDate = new Date("1992-06-26");

    // Set lastModified to 365 days ago
    const course = generateTestCourse({
      dateUpdated: testDate,
    });
    const component = renderer.create(
      <MemoryRouter>
        <LastEdited course={course} key={1} />
      </MemoryRouter>,
    );

    const pTag = component.root.findByType("p");
    const dateValue = pTag.children[0];

    expect(dateValue).toBe(testDate.toLocaleDateString());
  });
});
