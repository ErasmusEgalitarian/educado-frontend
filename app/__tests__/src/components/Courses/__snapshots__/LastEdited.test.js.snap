// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`last edited time shows days if difference is below 1 week 1`] = `
<div
  className="flex flex-row"
>
  <svg
    className="w-4 mr-1 inline-block mr-1"
    role="presentation"
    style={{}}
    viewBox="0 0 24 24"
  >
    <path
      d="M20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18,2.9 17.35,2.9 16.96,3.29L15.12,5.12L18.87,8.87M3,17.25V21H6.75L17.81,9.93L14.06,6.18L3,17.25Z"
      style={
        {
          "fill": "currentColor",
        }
      }
    />
  </svg>
  <p
    className="text-md italic"
  >
    6 dias atrás
  </p>
</div>
`;

exports[`last edited time shows hours if difference is below 24 hours 1`] = `
<div
  className="flex flex-row"
>
  <svg
    className="w-4 mr-1 inline-block mr-1"
    role="presentation"
    style={{}}
    viewBox="0 0 24 24"
  >
    <path
      d="M20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18,2.9 17.35,2.9 16.96,3.29L15.12,5.12L18.87,8.87M3,17.25V21H6.75L17.81,9.93L14.06,6.18L3,17.25Z"
      style={
        {
          "fill": "currentColor",
        }
      }
    />
  </svg>
  <p
    className="text-md italic"
  >
    23 horas atrás
  </p>
</div>
`;

exports[`last edited time shows minutes if difference is below 1 hour 1`] = `
<div
  className="flex flex-row"
>
  <svg
    className="w-4 mr-1 inline-block mr-1"
    role="presentation"
    style={{}}
    viewBox="0 0 24 24"
  >
    <path
      d="M20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18,2.9 17.35,2.9 16.96,3.29L15.12,5.12L18.87,8.87M3,17.25V21H6.75L17.81,9.93L14.06,6.18L3,17.25Z"
      style={
        {
          "fill": "currentColor",
        }
      }
    />
  </svg>
  <p
    className="text-md italic"
  >
    59 minutos atrás
  </p>
</div>
`;

exports[`last edited time shows months if difference is below 1 year 1`] = `
<div
  className="flex flex-row"
>
  <svg
    className="w-4 mr-1 inline-block mr-1"
    role="presentation"
    style={{}}
    viewBox="0 0 24 24"
  >
    <path
      d="M20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18,2.9 17.35,2.9 16.96,3.29L15.12,5.12L18.87,8.87M3,17.25V21H6.75L17.81,9.93L14.06,6.18L3,17.25Z"
      style={
        {
          "fill": "currentColor",
        }
      }
    />
  </svg>
  <p
    className="text-md italic"
  >
    12 meses atrás
  </p>
</div>
`;

exports[`last edited time shows weeks if difference is below 30 days 1`] = `
<div
  className="flex flex-row"
>
  <svg
    className="w-4 mr-1 inline-block mr-1"
    role="presentation"
    style={{}}
    viewBox="0 0 24 24"
  >
    <path
      d="M20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18,2.9 17.35,2.9 16.96,3.29L15.12,5.12L18.87,8.87M3,17.25V21H6.75L17.81,9.93L14.06,6.18L3,17.25Z"
      style={
        {
          "fill": "currentColor",
        }
      }
    />
  </svg>
  <p
    className="text-md italic"
  >
    29 semanas atrás
  </p>
</div>
`;
